### PR TITLE
Encrypt drafts to self

### PR DESCRIPTION
**Pull request by Valodim**
This was actually easier than it had any right to be. The loading logic was already there, requiring only very few changes to actually encrypt drafts. I think it's a bit problematic that this introduces potentially difficult error scenarios, but then again if the user's pgp setup isn't in order they probably don't want drafts saved in plaintext instead. Right now it's always possible to just disable encryption and save the draft in plaintext, we'll see how that goes.

I still need to test to ensure compatibility with at least enigmail. 

Encryption of subject will be added when protected-headers is implemented, a setting to encrypt all drafts (or don't encrypt but also don't upload?) might be in a follow-up PR.

### CODE REVIEW

**Comment by cketti** [comment]
Can we use the `\Draft` flag instead of adding another header? See #3291.

**Comment by Valodim** [comment]
the only part where this is even needed is https://github.com/k9mail/k-9/blob/master/k9mail/src/main/java/com/fsck/k9/controller/MessagingController.java#L2587 so sure, if we can use the draft flag there that's much better.

**Comment by Valodim** [comment]
This PR is as good as it gets, without saving the ui state in a header.

I added another fix that is only loosely related, which skips autocrypt-gossip and subject encryption for pgp/inline mails, since those rely on encryption of mime-headers.

**Comment by Valodim** [comment]
related: https://github.com/autocrypt/autocrypt/pull/376

**Comment by Valodim** [comment]
I tested this with Enigmail, and the autocrypt spec is likely to pass like this otherwise I'll adapt later. Good to merge from my side.

**Comment by Valodim** [comment]
rebased this branch, should be good now

File: app/core/src/main/java/com/fsck/k9/message/PgpMessageBuilder.java [hunk 1]
```diff
@@ -109,15 +111,27 @@ protected void buildMessageInternal() {
             return;
         }
 
+        addAutocryptHeaderIfAvailable(openPgpKeyId);
+        addDraftStateHeader();
```
Comment (cketti) [inline app/core/src/main/java/com/fsck/k9/message/PgpMessageBuilder.java]: Will this add the `Autocrypt-Draft-State` header to all messages, regardless of whether or not it's a draft message?

File: app/ui/src/main/java/com/fsck/k9/activity/compose/ComposeCryptoStatus.kt [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+import com.fsck.k9.message.CryptoStatus
+import com.fsck.k9.view.RecipientSelectView.Recipient
+import org.openintents.openpgp.OpenPgpApiManager
+import org.openintents.openpgp.OpenPgpApiManager.OpenPgpProviderState
+
+/** This is an immutable object which contains all relevant metadata entered
+ * during email composition to apply cryptographic operations before sending
+ * or saving as draft.
+ */
+data class ComposeCryptoStatus(private val openPgpProviderState: OpenPgpProviderState,
```
Comment (cketti) [inline app/ui/src/main/java/com/fsck/k9/activity/compose/ComposeCryptoStatus.kt]: Side note: For the Kotlin code style please see https://developer.android.com/kotlin/style-guide and https://kotlinlang.org/docs/reference/coding-conventions.html#class-header-formatting

Comment (cketti) [inline app/ui/src/main/java/com/fsck/k9/activity/compose/ComposeCryptoStatus.kt]: More of a general comment: If there's a lot of code in a data class maybe that's a sign that the logic should be extracted to a separate class and only the output that is ready for consumption should be put into a data class.

File: app/core/src/main/java/com/fsck/k9/preferences/AccountSettings.java [hunk 3]
```diff
@@ -239,6 +239,9 @@
         s.put("openPgpEncryptSubject", Settings.versions(
                 new V(51, new BooleanSetting(true))
         ));
+        s.put("openPgpEncryptAllDrafts", Settings.versions(
+                new V(52, new BooleanSetting(true))
```
Comment (cketti) [inline app/core/src/main/java/com/fsck/k9/preferences/AccountSettings.java]: Looks like you forgot to increment `Settings.VERSION`.

File: app/core/src/main/java/com/fsck/k9/message/PgpMessageBuilder.java (hunk 1 above)
Comment (Valodim) [reply app/core/src/main/java/com/fsck/k9/message/PgpMessageBuilder.java]: it did. weird, I had it right in an earlier commit. :)

**Comment by Valodim** [comment]
Did some minor cleaning, and added unit tests. Please merge after quick sanity check :+1:

File: app/ui/src/main/java/com/fsck/k9/activity/MessageCompose.java [hunk 4]
```diff
@@ -1325,6 +1328,9 @@ private void processDraftMessage(MessageViewInfo messageViewInfo) {
         // See buildIdentityHeader(TextBody) for a detailed description of the composition of this blob.
         Map<IdentityField, String> k9identity = new HashMap<>();
         String[] identityHeaders = message.getHeader(K9.IDENTITY_HEADER);
+        if (identityHeaders.length == 0) {
+            identityHeaders = messageViewInfo.rootPart.getHeader(K9.IDENTITY_HEADER);
```
Comment (cketti) [inline app/ui/src/main/java/com/fsck/k9/activity/MessageCompose.java]: Wouldn't it be better to use the inner header and only fall back to the outer one?

Comment (Valodim) [reply app/ui/src/main/java/com/fsck/k9/activity/MessageCompose.java]: Yes, slightly :)

**Pull request merged**